### PR TITLE
Actually Run Tests in CI

### PR DESCRIPTION
**Pull request by jamiebrynes7**
#### Description
TL;DR - Adding `Generate` as a `delayCall` on startup means that when CI runs, the GDK assemblies actually fail to build - but this doesn't cause Unity to exit with a non-zero exit code 😭. It will silently ignore the compilation failure and run tests from any successfully built assemblies.

This PR reverts the that change introduced in #286. This required moving the "GDK Tools Configuration" from a `ScriptableObject` to a raw JSON file - as the `AssetDatabase` is not available during compilation.

Swapped the custom inspector view for the `ScriptableObject` to an Editor window.
#### Tests
Ran `ci/test.sh` from a clean state and made sure that tests were actually picked up.
#### Documentation
N/A
#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.
@jessicafalk @jared-improbable

### CODE REVIEW

File: workers/unity/Packages/com.improbable.gdk.tools/GdkToolsConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+            {
+                value = castedValue;
+                return true;
+            }
+
+            value = default(TValue);
+            return false;
+        }
+
+        private Dictionary<string, object> ToJson()
```
Comment (zeroZshadow) [inline workers/unity/Packages/com.improbable.gdk.tools/GdkToolsConfiguration.cs]: This doesn't really make JSON

File: workers/unity/Packages/com.improbable.gdk.tools/GdkToolsConfiguration.cs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+            return data;
+        }
+
+        public static GdkToolsConfiguration GetOrCreateInstance()
+        {
+            return File.Exists(JsonFilePath) ? LoadFile() : CreateInstance();
+        }
+
+        private static GdkToolsConfiguration LoadFile()
```
Comment (zeroZshadow) [inline workers/unity/Packages/com.improbable.gdk.tools/GdkToolsConfiguration.cs]: `LoadFromFile`

File: workers/unity/ProjectSettings/EditorSettings.asset [hunk 3]
```diff
@@ -14,8 +14,9 @@ EditorSettings:
   m_EtcTextureFastCompressor: 1
   m_EtcTextureNormalCompressor: 2
   m_EtcTextureBestCompressor: 4
-  m_ProjectGenerationIncludedExtensions: txt;xml;fnt;cd
+  m_ProjectGenerationIncludedExtensions: txt;xml;fnt;cd;asmdef
```
Comment (jessicafalk) [inline workers/unity/ProjectSettings/EditorSettings.asset]: do we need that change?

File: test-project/Assets/Config/GdkToolsConfiguration.json [hunk 4]
```diff
@@ -0,0 +1 @@
+{"schema_standard_lib_directory":"build/dependencies/schema/standard_library","schema_source_directories":["Assets/.Schema"],"codegen_output_directory":"Assets/Generated/Source"}
```
Comment (firtina-improbable) [inline test-project/Assets/Config/GdkToolsConfiguration.json]: nit: can you put fields in separate lines (will help with reviewing when parts change later)

File: workers/unity/ProjectSettings/EditorSettings.asset (hunk 3 above)
Comment (firtina-improbable) [reply workers/unity/ProjectSettings/EditorSettings.asset]: I thought we should have committed this in already- I remember overhearing people talking about this file. I believe it comes from when we last upgraded Unity version, whenever you save the project it'd appear again.

File: test-project/Assets/Config/GdkToolsConfiguration.json (hunk 4 above)
Comment (jamiebrynes7) [reply test-project/Assets/Config/GdkToolsConfiguration.json]: I can manually make the change - but any auto-generated JSON will have this problem 😢

Comment (firtina-improbable) [reply test-project/Assets/Config/GdkToolsConfiguration.json]: `JsonUtility` has pretty-printing capabilities - it should also make it easier to parse into/out of a `[Serializable]` struct rather than a dictionary. It may be worth giving it a try if you haven't already :)

Comment (jamiebrynes7) [reply test-project/Assets/Config/GdkToolsConfiguration.json]: The reason I didn't use `JsonUtility` is that it doesn't support lists/arrays out of the box and you have to implement wrappers to work around it.

We also use `MiniJSON` elsewhere so was following precedent. I think @jared-improbable has some oppositions to `JsonUtility`

Comment (firtina-improbable) [reply test-project/Assets/Config/GdkToolsConfiguration.json]: Lists/arrays work, dictionaries are the ones that are difficult.

```cs
        [Serializable]
        public struct SomeStruct
        {
            public string schema_standard_lib_directory;
            public string[] schema_source_directories;
            public string codegen_output_directory;
        }

        [Test]
        public void JSONUtilityTest()
        {
            var myStruct = JsonUtility.FromJson<SomeStruct>(@"{ ""schema_standard_lib_directory"":""build/dependencies/schema/standard_library"", ""schema_source_directories"":[""Assets/.Schema""],""codegen_output_directory"":""Assets/Generated/Source""}");

            Debug.Log(JsonUtility.ToJson(myStruct, true));
        }
```

Output:
```json
{
    "schema_standard_lib_directory": "build/dependencies/schema/standard_library",
    "schema_source_directories": [
        "Assets/.Schema"
    ],
    "codegen_output_directory": "Assets/Generated/Source"
}
```

Comment (firtina-improbable) [reply test-project/Assets/Config/GdkToolsConfiguration.json]: It's Sunday so I'll allow myself a meme

internet-lies

Comment (jamiebrynes7) [reply test-project/Assets/Config/GdkToolsConfiguration.json]: Adjusted to use `JSONUtility`, PTAL :)

File: workers/unity/Packages/com.improbable.gdk.tools/GdkToolsConfigurationEditorExtensions.cs [hunk 5]
```diff
@@ -1,16 +1,14 @@
...
-using Object = UnityEngine.Object;
 
 namespace Improbable.Gdk.Tools
 {
     /// <summary>
     ///     Defines a custom inspector window that allows you to configure the GDK Tools.
     /// </summary>
-    [CustomEditor(typeof(ScriptableGdkToolsConfiguration))]
-    public class GdkToolsConfigurationInspector : Editor
+    public class GdkToolsConfigurationWindow : EditorWindow
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/GdkToolsConfigurationEditorExtensions.cs]: Potential for improvement: wondering if we should do a prompt to save file on disable of this window, or save at any modification? I like the latter more, fewer steps. The user can get unexpected results if they change the fields here but forget to hit the save button. The remaining problem is lack of undo - but that's something harder to implement for now.

File: workers/unity/Packages/com.improbable.gdk.tools/GdkToolsConfigurationEditorExtensions.cs [hunk 6]
```diff
@@ -55,19 +59,19 @@ public override void OnInspectorGUI()
                 DrawCodeGenerationOptions();
                 DrawHorizontalBreak();
 
+                GUI.enabled = configErrors.Count == 0;
```
[the remark this answers is not in the document]
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/GdkToolsConfigurationEditorExtensions.cs]: Nice - will use that

File: workers/unity/Packages/com.improbable.gdk.tools/GdkToolsConfigurationEditorExtensions.cs (hunk 5 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/GdkToolsConfigurationEditorExtensions.cs]: Probably don't want to save at any modification - as you can get into a bad state unintentionally. I prefer the prompt myself. 

We can look at this in a future PR as I'd like to get this in sooner rather than later so tests run in CI again 😬

Comment (jessicafalk) [reply workers/unity/Packages/com.improbable.gdk.tools/GdkToolsConfigurationEditorExtensions.cs]: agreed, for now we should get this in asap to get tests running again

File: workers/unity/ProjectSettings/EditorSettings.asset [hunk 7]
```diff
@@ -14,8 +14,9 @@ EditorSettings:
...
   m_ProjectGenerationRootNamespace: 
   m_UserGeneratedProjectSuffix: 
   m_CollabEditorSettings:
     inProgressEnabled: 1
+  m_EnableTextureStreamingInPlayMode: 1
```
Comment (jared-improbable) [inline workers/unity/ProjectSettings/EditorSettings.asset]: ?

Comment (jamiebrynes7) [reply workers/unity/ProjectSettings/EditorSettings.asset]: We wanted this line - `m_ProjectGenerationIncludedExtensions: txt;xml;fnt;cd;asmdef` 

But this one snuck in as well - we can remove? But it always shows up in diffs

Comment (jared-improbable) [reply workers/unity/ProjectSettings/EditorSettings.asset]: I guess if Unity really wants that setting, may as well let it have its way.
